test/test_optional.cpp

```diff
@@ -0,0 +1,697 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sparrow/optional.hpp"
+
+#include "doctest/doctest.h"
+
+namespace sparrow
+{
+    using optional_double = optional<double, bool>;
```

Comment:
Missing tests I would like to see:
- find a way to have all the tests in `consteval` context, to make sure the `constexpr` in the class definition all work compile-time (except maybe for the custom type below); Note that code that can be tested at compile-time is guaranteed to not compile if there is UB situations in it, so that's a big win.
- parameterize the tests so that we can check with other type than `double` and `int`
- set T as a custom type which set a global boolean on construction and destruction
  - it must not be set instantiated when default-construction the optional
  - it must be destroyed when the boolean is set to false